src/charm.py

```diff
@@ -28,9 +30,91 @@ def __init__(self, *args: typing.Any):
         """
         super().__init__(*args)
         self.chrony = Chrony()
+        self.certificates = tls_certificates.TLSCertificatesRequiresV3(self, "certificates")
+        self.tls_keychain = TlsKeychain(charm=self, namespace="certificates")
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.certificates_relation_created, self._on_certificates_relation_created
+        )
+        self.framework.observe(
+            self.on.certificates_relation_created, self._on_certificate_unavailable
+        )
+        self.framework.observe(
+            self.on.certificates_relation_broken, self._on_certificates_relation_broken
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_unavailable
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated, self._on_certificate_unavailable
+        )
+
+    def _on_certificates_relation_created(self, _: ops.RelationCreatedEvent) -> None:
+        """Handle the certificates relation-creation event."""
+        if self.unit.is_leader():
+            self.unit.open_port("tcp", 4460)
+
+    def _on_certificates_relation_broken(self, _: ops.RelationBrokenEvent) -> None:
+        """Handle the certificates relation-broken event."""
+        if self.unit.is_leader():
+            self.unit.close_port("tcp", 4460)
+        self.tls_keychain.clear()
+        self._do_config()
+
+    def _on_certificate_unavailable(self, _: ops.CharmEvents) -> None:
+        """Handle the event when certificate is unavailable."""
+        if not self._get_server_name():
+            return
+        self._renew_certificate()
+
+    def _on_certificate_available(self, event: tls_certificates.CertificateAvailableEvent) -> None:
+        """Handle the certificate-available event.
+
+        Args:
+            event: certificate-available event object.
+        """
+        self.tls_keychain.set_chain(event.chain_as_pem())
+        self._do_config()
+
+    def _renew_certificate(self) -> None:
+        """Renew the certificate.
+
+        Raises:
+            RuntimeError: canary exception
+        """
+        old_csr = self.tls_keychain.get_csr()
+        private_key = self.tls_keychain.get_private_key()
+        if not self._get_server_name():
+            # canary exception
```

Comment:
Yes, if this exception is raised, that means I did something wrong.

Reply:
Use of `assert` will trigger a bandit warning, which needs to be masked by `nosec`. Would you prefer that approach? I can update it to use `assert`.

Reply:
Indeed, that's the second time we get this questioning, I'm in favor of using `assert` for "impossible" cases, despite the bandit warning.
`assert` has a drawback of not being executed in normal conditions, but these cases are mostly there as preconditions (or postconditions) than data validation, so I think it's fine to not have them when the charm is running in normal conditions.
@jdkandersson WDYT?

Reply:
I changed to `assert` + `nosec` for now, thanks!

Reply:
I am not so much in favour of this, but yes I asked if I should use it and included it due to https://github.com/canonical/github-runner-webhook-router/pull/11#discussion_r1656551028.

I would prefer to have an
```
if condition not true:  # pragma: no cover
   raise AssertionError
```
as this will also work for optimised code.

Reply:
Yes, I think that's a very good idea, updated to use `AssertionError`, thanks!